src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -16,79 +17,104 @@ void MPIDI_OFI_deferred_am_isend_dequeue(MPIDI_OFI_deferred_am_isend_req_t * dre
 {
     int mpi_errno = MPI_SUCCESS;
-    int dt_contig = 0;
-    MPI_Aint data_sz;
-    MPI_Aint dt_true_lb, last;
-    MPIR_Datatype *dt_ptr;
     char *send_buf;
-    bool need_packing = false;
-    bool do_eager = false;
-
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
+    MPI_Aint packed_size;
+    MPI_Aint dt_true_lb, send_size, offset;
+    MPIR_Datatype *dt_ptr;
+    MPIDI_OFI_am_send_request_t *send_req;
 
-    MPIDI_Datatype_get_info(dreq->count, dreq->datatype, dt_contig, data_sz, dt_ptr, dt_true_lb);
-    need_packing = dt_contig ? false : true;
-    send_buf = (char *) dreq->buf + dt_true_lb;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_ISSUE_DEFERRED_AM_ISEND);
 
-    do_eager =
-        (MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, am_hdr_sz) + dreq->data_sz
-         + sizeof(MPIDI_OFI_am_header_t) <= MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE);
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "try issue deferred am isend dreq=%p", dreq));
 
-    MPL_pointer_attr_t attr;
-    MPIR_GPU_query_pointer_attr(dreq->buf, &attr);
-    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
-        /* Force packing of GPU buffer in host memory */
-        need_packing = true;
+    if (dreq->is_pipeline) {
+        offset = MPIDIG_send_get_offset(dreq->sreq);
+        send_size = MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE - (offset ? 0 : dreq->am_hdr_sz)
+            - sizeof(MPIDI_OFI_am_header_t);
+        send_size = MPL_MIN(send_size, MPIDIG_send_get_data_sz_left(dreq->sreq));
+    } else {
+        send_size = dreq->data_sz;
+        offset = 0;
     }
 
-    if (need_packing) {
-        /* FIXME: currently we always do packing, also for high density types. However,
-         * we should not do packing unless needed. Also, for large low-density types
-         * we should not allocate the entire buffer and do the packing at once. */
-        /* TODO: (1) Skip packing for high-density datatypes;
-         *       (2) Pipeline allocation for low-density datatypes; */
-        if (do_eager) {
-            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
-                                                           (void **) &send_buf);
-            if (send_buf == NULL) {
-                /* no buffer available, suppress error and exit */
-                mpi_errno = MPI_SUCCESS;
-                goto fn_exit;
-            }
-            MPIR_ERR_CHECK(mpi_errno);
-        } else {
-            /* only for large message RDMA read */
-            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
+    if (dreq->need_packing) {
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
```

Comment:
We need always check `mpi_errno`. Does `send_buf==NULL` still return `MPI_SUCCESS`?

Reply:
We rarely capture the return code for genq operations. Is there any special reason to do so here?

Reply:
It is for code cleanness. If we always intend to ignore error return, then we should modify the function to return `void`.

Reply:
I agree. I'm just surprised to see the error captured and then masked here. In current `main` I see 14 instances of `MPIDU_genq_private_pool_alloc_cell`. Only 1 captures the return code. If the function can fail, then we should always check. IMO, returning `NULL` is not an "error" and is already handled separately.

Reply:
This still need address.

Reply:
ok

Reply:
I missed this comment.